text/0000-target-tier-policy.md

```diff
@@ -0,0 +1,222 @@
+
+## Tier 1 target policy
+
+At this tier, the Rust project guarantees that a target builds and passes all
+tests, and will reject patches that fail to build or pass the testsuite on a
+target. We hold tier 1 targets to our highest standard of requirements.
+
+- Any new tier 1 target requires compiler team approval based on these
+  requirements.
+- In addition, the infrastructure team must approve the integration of the
+  target into CI, and the CI-related requirements. This review and approval
+  will typically take place in the PR adding the target to CI.
+- In addition, the release team must approve the long-term viability of the
+  target, and the additional work of supporting the target.
+- Tier 1 targets must have substantial, widespread interest within the
+  developer community, and must serve the ongoing needs of multiple production
+  users of Rust across multiple organizations or projects. A tier 1 target may
+  be demoted or removed if it becomes obsolete or no longer meets this
+  requirement.
+- The target must build and pass tests reliably in CI.
+- Building the target and running the testsuite for the target must not take
+  substantially longer than other targets.
+- If running the testsuite requires additional infrastructure (such as systems
+  running the target), the target development team shall arrange to provide
+  such resources to the Rust project, to the satisfaction and approval of the
+  Rust infrastructure team.
+- Tier 1 targets must provide documentation for the Rust community for how to
+  build and run tests for the platform, using emulation if possible, or
+  dedicated hardware if necessary.
+- A tier 1 target may be demoted or removed if it no longer meets these
+  requirements. Any proposal for demotion or removal will be communicated
+  widely to the Rust community, both when initially proposed and before being
+  dropped from a stable release.
+- All tier 2 requirements apply.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+While these criteria attempt to document the policy, that policy still involves
+human judgment. Targets must fulfill the spirit of the requirements as well, as
+determined by the judgment of the approving teams.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The set of approving teams for each tier arose out of discussion with the
+various teams involved with aspects of the Rust project impacted by new
+targets.
+
+
+# Prior art
+[prior-art]: #prior-art
+
+This attempts to formalize and document Rust policy around targets and
+architectures. https://forge.rust-lang.org/release/platform-support.html
+documents some of that policy.
+
+Future expansions of such policy may find requirements from other communities
+useful as examples, such as Debian's [arch
+policy](https://release.debian.org/bullseye/arch_policy.html) and [archive
+criteria](https://ftp-master.debian.org/archive-criteria.html).
+
+# Unresolved questions
```

Comment:
This document seems to aim to give high level guidance vs specifying specific procedure when it comes to who exactly is responsible for the final decision for promoting and demoting a target to different tiers and how exactly that decision should be made. Should there be specific guidance as to who gets ultimate say if there is open disagreement with a decision? Is the specific policy of how such a decision is made left up to the discretion of the responsible team and is this document meant to be the only guidance to those teams for how they should make those decisions?

Reply:
@rylev That is an interesting question, since to my knowledge there haven't been any intentional promotions or demotions of a platform. To me platform promotions are a core team decision as it involves coordinating with multiple teams to provide the stability to the target.

Reply:
The way I see it, each of the three teams involved in this (compiler, infra, release) has a clearly defined stake in the decision, and if any of them is not comfortable promoting the target they should be able to veto the promotion:

* If the compiler team can't ensure "tier 1 quality" for the target and nobody is available to do the development work there is little point in promoting the target.
* If the infrastructure team can't provide CI for the platform the only guarantee we give for tier 1 platforms (they're tested on CI) goes out of the window.
* If the release team doesn't think the platform is viable in the long term or that shipping releases for it is going to be too cumbersome why are we doing the promotion?

I don't see how the core team needs to get involved in the decision.

Reply:
@pietroalbini I would call a decision made between the compiler, infra, and release leads a core team decision, and I don't see why it couldn't involve other team's if they had concerns, and we already have the core team meetings which are recorded and have the public portion published.

Reply:
The leads are not the teams tho. That's a difference between like 3 and 30 people. We currently don't give special significance to team leaders this way and I don't see why we would start with a decision of such importance as tier 1 status.

Reply:
No one is saying that three people are making the full decision, however there won't be a RFC that requires 30 people to approve, the point of a team lead is to represent their team and its members, it doesn't require everyone to be personally involved. [RFC 1068] states that the core team _"Sets the priorities and release schedule"_, and _"Focuses on broad, cross-cutting concerns"_. Now that document is quite old, however I don't think that section has changed much, and to me platform target support falls under both those responsibilities.

[rfc 1068]: https://github.com/rust-lang/rfcs/blob/master/text/1068-rust-governance.md#core-team

Reply:
> No one is saying that three people are making the full decision, however there won't be a RFC that requires 30 people to approve.

I don't see anything wrong with 30 people (probably less as some people are on several teams) approving a decision as big as adding a new tier 1 target. It's not like we do that every day, in fact, it hasn't happened for years.

> The point of a team lead is to represent their team and its members, it doesn't require everyone to be personally involved.

I think we can let the team decide, if they want, to delegate to someone on the team if it is a regular task that needs delegation for time-resource reasons. As for tier 3 targets, they are added regularly by anyone familiar with the compiler internals of adding target on a regular basis; it doesn't even require approval from any team.

> [RFC 1068](https://github.com/rust-lang/rfcs/blob/master/text/1068-rust-governance.md#core-team) states that the core team _"Sets the priorities and release schedule"_, and _"Focuses on broad, cross-cutting concerns"_. Now that document is quite old, however I don't think that section has changed much, and to me platform target support falls under both those responsibilities.

I don't think it does; The domain expertise required for making decisions about targets mainly resides in the compiler, infra, and release teams. It may not reside in the team leads of those teams but it should in the collective knowledge of each whole team.

Reply:
> I don't think it does; The domain expertise required for making decisions about targets mainly resides in the compiler, infra, and release teams. It may not reside in the team leads of those teams but it should in the collective knowledge of each whole team.

The platform support promotion isn't solely a technical decision. Tiers are a form of guarantee not a statement of intent. To discuss whether a platform should be promoted we'd have to already have the compiler support and release infrastructure required in place and functioning. The technical requirements are a prerequisite to even having the discussion of promotion.

Reply:
Judging whether we have e.g. the CI budget or whether e.g. whether there aren't any `unimplemented!()`s are indeed some sort of technical requirements. However, the purpose of team review here is to make sure that problems don't slip through the cracks. Having the watchful eye of many people rather than just 3 team leads, who are typically some of the most busy people on the team, makes it more likely that things don't slip through.

Also, I don't really understand why non-technical decisions would better be made by leads rather than by the whole team (again, within the discretion of the team to delegate, sometimes to someone who isn't the lead).

Reply:
@Centril Sorry I don't think I'm being clear. What I'm trying to say is that in order to get to the point where we could have a discussion about promoting a platform, those technical problems would have to already be solved. For example; we wouldn't make the decision to promote `x86_64-unknown-redox` from tier 2 to tier 1 and then try to make that platform tier 1 quality. That would be putting the cart before the horse. To me a platform should already be de-facto tier 1 before we would commit to making that happen, and that is something that part is something that is up to the teams to agree to work together to provide that.

Reply:
@XAMPPRocky Oh sure; that seems right, but the function of team review would be to do a *final verification* that it is indeed the case that a tier 2 platform satisfies the wrt. the technical requirements of libstd impl maturity, CI budget, maintainers, etc. are met for tier 1 and can be promoted.

Reply:
I've added some discussion about how alternatives for handling and coordinating team consensus.